Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py

```diff
@@ -6458,6 +6695,907 @@ def get_list_asset_group_command(
+                # The .strip() is REMOVED to preserve any trailing whitespace/newlines
+                # for the CSV parser (though newlines are usually handled by splitlines).
+                yield line_bytes.decode("utf-8")
+
+        # Process any final content left in the buffer
+        if buffer:
+            yield buffer.decode("utf-8")
+
+        demisto.debug("Finished streaming report.")
+
+    finally:
+        # Crucial: Always ensure the response object is released/closed
+        await response.release()
+
+
+async def download_and_parse_report(
+    client: "InsightVMClient",
+    report_id: str,
+    instance_id: str,
+    event_integration_context: dict,
+    event_type: str,
+    batch_size: int = 1000,
+):
+    # --- Checkpoint Setup ---
+    current_line_count = 0
+    # Retrieve checkpoints:
+    snapshot_id = event_integration_context.get("snapshot_id", str(round(time.time() * 1000)))
+    last_sent_line_raw = event_integration_context.get("last_sent_line", 0)
+    total_records_ingested = event_integration_context.get("total_records_ingested", 0)
+
+    current_batch: List[str] = []
+    header: Optional[List[str]] = None
+
+    HEADER_LINE_NUMBER = 1
+    start_line_for_batch_raw = last_sent_line_raw + 1
+    pending_tasks = set()
+    batch_counter = 0  # count the number of batches
+
+    demisto.debug(f"--- Starting Data Stream (Resuming from line: {start_line_for_batch_raw}) ---")
+    # --- Lookahead Stream Setup ---
+    stream_iterator = stream_report(client, report_id, instance_id).__aiter__()
+    line_to_process: Optional[str] = None
+
+    # Track data records processed in the current run (resets to 0 on crash/restart)
+    current_run_data_records = 0
+
+    demisto.debug("--- Starting Data Stream (Final Count Calculated On-The-Fly) ---")
+
+    try:
+        # 1. INITIAL LINE READ: Get the very first line of the stream
+        try:
+            line_to_process = await anext(stream_iterator)
+        except StopAsyncIteration:
+            demisto.debug("Report stream was empty.")
+            return  # Exit if the report is empty
+
+        while line_to_process is not None:
+            current_line_count += 1
+            current_line = line_to_process
+
+            # --- Try to read the NEXT line (Lookahead) ---
+            try:
+                next_line = await anext(stream_iterator)
+            except StopAsyncIteration:
+                # If we fail to get the next line, this is the final line of the report.
+                next_line = None
+
+            # 1a. Process Header (Line 1)
+            if current_line_count == HEADER_LINE_NUMBER:
+                try:
+                    header = next(csv.reader(io.StringIO(current_line)))
+                    line_to_process = next_line
+                    continue
+                except Exception as e:
+                    raise Exception(f"Failed to parse CSV header on line {current_line_count}: {e}")
```

Comment:
Let's consider taking this condition out of the while to avoid this check every iteration.